### PR TITLE
runfiles: port phst/runfiles to rules_go

### LINKED ISSUE #2359: runfiles.go should look at os.Args[0] + ".runfiles" etc.
### What version of rules_go are you using?

0.21.2

### What version of gazelle are you using?

None

### What version of Bazel are you using?

2.0.0

### Does this issue reproduce with the latest releases of all the above?

Yes

### What operating system and processor architecture are you using?

Debian GNU/Linux, x86-64

### Any other potentially useful information about your toolchain?

No

### What did you do?

I created the following workspace:

- BUILD:

```python
load("@io_bazel_rules_go//go:def.bzl", "go_binary", "go_library", "go_test")

go_binary(
    name = "test",
    srcs = ["test.go"],
    data = ["test.txt"],
    deps = ["@io_bazel_rules_go//go/tools/bazel:go_default_library"],
)

genrule(
    name = "genrule",
    outs = ["copy.txt"],
    cmd = "$(location :test) > $@",
    tools = [":test"],
)
```

- test.go:

```go
package main

import (
	"io"
	"log"
	"os"

	"github.com/bazelbuild/rules_go/go/tools/bazel"
)

func main() {
	filename, err := bazel.Runfile("test.txt")
	if err != nil {
		log.Fatal(err)
	}
	fd, err := os.Open(filename)
	if err != nil {
		log.Fatal(err)
	}
	defer fd.Close()
	if _, err := io.Copy(os.Stdout, fd); err != nil {
		log.Fatal(err)
	}
}
```

- test.txt with arbitrary contents.

Then run `bazel build :copy.txt`.

### What did you expect to see?

Command succeeds

### What did you see instead?

```none
could not locate runfiles directory
Target //:copy.txt failed to build
```

* * * *

The issue is that Bazel actions don't pass runfiles environment variables, and the current directory is also incorrect. Rather, runfiles.go should look at `os.Args[0] + ".runfiles"`, `os.Args[0] + ".runfiles/MANIFEST"`, and `os.Args[0] + ".runfiles_manifest"`. See the C++ code for comparison: https://github.com/bazelbuild/bazel/blob/master/tools/cpp/runfiles/runfiles_src.cc#L271-L280.

### LINKED ISSUE #2422: Runfiles should be relative to runfiles root, not workspace subdirectories
Currently runfiles.go searches the workspace subdirectories for runfiles (https://github.com/bazelbuild/rules_go/blob/master/go/tools/bazel/runfiles.go#L58). That's inconsistent with all other runfile implementations I'm aware of, and shouldn't be needed, as users are in practice always aware of the workspace they want to use. I think this functionality should be removed, i.e. runfiles should always be relative to the runfiles root, like in the other languages.

This would be an incompatible change, so probably this should be implemented as a new function/library, with the current ones deprecated.

### LINKED ISSUE #2423: Runfiles library should provide environment variables for subprocesses
If a Bazel-built binary wants to launch another Bazel-built binary, it needs to tell the latter binary where to find the runfiles, because `argv[0]` for the latter binary will be different (or, in case of scripts, entirely unavailable). Other languages provide a function that returns `RUNFILES_DIR` and `RUNFILES_MANIFEST` environment variables (see https://github.com/bazelbuild/bazel/blob/master/tools/cpp/runfiles/runfiles_src.h#L159 for C++), and runfiles.go should provide the same:

```go
func EnvVars() []string
```

### LINKED ISSUE #3036: Upstream `github.com/phst/runfiles` as `@io_bazel_rules_go//go/runfiles`
rules_go currently doesn't offer a way to discover runfiles following the [standard procedures for runfiles discovery](https://docs.google.com/document/d/e/2PACX-1vSDIrFnFvEYhKsCMdGdD40wZRBX3m3aZ5HhVj4CtHPmiXKDCxioTUbYsDydjKtFDAzER5eg7OjJWs3V/pub). There is https://pkg.go.dev/github.com/bazelbuild/rules_go/go/tools/bazel, but due to it predating the coordinated efforts to provide consistent runfiles libraries for all languages, it uses a non-standard procedure and doesn't interoperate well with the other libraries. It also invokes a bit too much magic that makes it prone to unexpected issues (e.g. it searches the working directory first before looking up the runfiles variables).

A proper replacement has been available as https://pkg.go.dev/github.com/phst/runfiles for quite some time now, so I would propose to upstream it (and would offer to do the work). According to the ["Deploying rules"](https://docs.bazel.build/versions/main/skylark/deploying.html#runfiles-library) guide, this runfiles library should be available under `@io_bazel_rules_go//go/runfiles` (Gazelle could probably be taught to pick it up automatically).

### PR DESCRIPTION
**Pull request by sluongng**
Today rules_go provide //go/tools/bazel as the canonical runfiles
library for binaries to be executed with `bazel test` and `bazel run`.

However, the current implementation pre-date the recent changes in
Bazel's upstream.  Since then, all of the native runfiles library of
Bash, Java, CPP, Python have been refactored to follow a certain
convention in locating files. (1)

Although these are subjected to change with the incoming BzlMod feature,
it would be easier to maintain if we can keep rules_go's runfiles
library implementation aligned with native languages' implementation.

Today, it seems like https://github.com/phst/runfiles implemented
exactly that.  So with @fmeum suggestion and @phst permission (2), let's
port the newer, more accurate implementation to rules_go.

Future refactoring will mark the current exported APIs in
//go/tools/bazel as deprecated and/or swapping out the old implementation
underneath to use this newer package.

Changes in this PR included:
- Copy paste repository over
- Removal of .git and .gitignore and .githooks dir
- Removal of repository specific files: README, WORKSPACE, LICENSE, CONTRIBUTE etc...
- Rename BUILD to BUILD.bazel
- Rename import path for both go and BUILD files
- Run gazelle over the packages
- Adjusted test cases to reflect new package paths
- Removed godoc related to installation instruction

(1): https://docs.google.com/document/d/e/2PACX-1vSDIrFnFvEYhKsCMdGdD40wZRBX3m3aZ5HhVj4CtHPmiXKDCxioTUbYsDydjKtFDAzER5eg7OjJWs3V/pub
(2): https://github.com/phst/runfiles/issues/3#issuecomment-1003070026

Close #3036 
Close #2359 
Close #2422 
Close #2423

### CODE REVIEW

**Comment by sluongng** [comment]
cc: @fmeum and @phst 🙏

**Comment by sluongng** [comment]
```
Executing tests from //go/tools/bazel/runfiles:runfiles_test
-----------------------------------------------------------------------------
--- FAIL: TestFS (0.00s)
    fs_test.go:41: TestFS found errors:
        .: Open: open .: file does not exist
        expected but not found: io_bazel_rules_go/go/tools/bazel/runfiles/test.txt
        expected but not found: io_bazel_rules_go/go/tools/bazel/runfiles/testprog/testprog
--- FAIL: TestRunfiles_manifestWithDir (0.00s)
    --- FAIL: TestRunfiles_manifestWithDir/foo/dir (0.00s)
        runfiles_test.go:126: Path failed: got "path\\to\\foo\\dir", want "path/to/foo/dir"
    --- FAIL: TestRunfiles_manifestWithDir/foo/dir/file (0.00s)
        runfiles_test.go:126: Path failed: got "path\\to\\foo\\dir/file", want "path/to/foo/dir/file"
    --- FAIL: TestRunfiles_manifestWithDir/foo/dir/deeply/nested/file (0.00s)
        runfiles_test.go:126: Path failed: got "path\\to\\foo\\dir/deeply/nested/file", want "path/to/foo/dir/deeply/nested/file"
--- FAIL: ExampleRunfiles (0.00s)
panic: exec: "C:\\b\\osvue4qe\\execroot\\io_bazel_rules_go\\bazel-out\\x64_windows-fastbuild\\bin\\go\\tools\\bazel\\runfiles\\testprog\\testprog": file does not exist [recovered]
	panic: exec: "C:\\b\\osvue4qe\\execroot\\io_bazel_rules_go\\bazel-out\\x64_windows-fastbuild\\bin\\go\\tools\\bazel\\runfiles\\testprog\\testprog": file does not exist

goroutine 1 [running]:
testing.(*InternalExample).processRunResult(0xc0000c7c70, {0x0, 0x0}, 0xc00007aae0, 0x0, {0x56f340, 0xc0001060e0})
	GOROOT/src/testing/example.go:91 +0x505
testing.runExample.func2()
	GOROOT/src/testing/run_example.go:60 +0x11c
panic({0x56f340, 0xc0001060e0})
	GOROOT/src/runtime/panic.go:1038 +0x215
go/tools/bazel/runfiles/runfiles_test_test.ExampleRunfiles()
	go/tools/bazel/runfiles/runfiles_test.go:59 +0x1d3
testing.runExample({{0x592d02, 0xf}, 0x59ed70, {0x5903e4, 0x4}, 0x0})
	GOROOT/src/testing/run_example.go:64 +0x28d
testing.runExamples(0xc0000c7e38, {0x6bf0a0, 0x2, 0x6})
	GOROOT/src/testing/example.go:44 +0x18e
testing.(*M).Run(0xc0000ec180)
	GOROOT/src/testing/testing.go:1505 +0x587
main.main()
	bazel-out/x64_windows-fastbuild/bin/go/tools/bazel/runfiles/runfiles_test_/testmain.go:110 +0x2b4
```

Ah window 🙃

**Comment by sluongng** [comment]
```
> git diff HEAD@{2}..HEAD
diff --git a/go/tools/bazel/runfiles/runfiles_test.go b/go/tools/bazel/runfiles/runfiles_test.go
index 610ca79f..78fb1ce7 100644
--- a/go/tools/bazel/runfiles/runfiles_test.go
+++ b/go/tools/bazel/runfiles/runfiles_test.go
@@ -47,7 +47,11 @@ func ExampleRunfiles() {
        }
        // The binary “testprog” is itself built with Bazel, and needs
        // runfiles.
-       prog, err := r.Path("io_bazel_rules_go/go/tools/bazel/runfiles/testprog/testprog")
+       rlocation := "io_bazel_rules_go/go/tools/bazel/runfiles/testprog/testprog"
+       if runtime.GOOS == "windows" {
+               rlocation = strings.ReplaceAll(rlocation, "/", "\\")
+       }
+       prog, err := r.Path(rlocation)
        if err != nil {
                panic(err)
        }
@@ -121,6 +125,7 @@ func TestRunfiles_manifestWithDir(t *testing.T) {
        } {
                t.Run(rlocation, func(t *testing.T) {
                        if runtime.GOOS == "windows" {
+                               rlocation = strings.ReplaceAll(rlocation, "/", "\\")
                                want = strings.ReplaceAll(want, "/", "\\")
                        }
                        got, err := r.Path(rlocation)
```

Tried some cheap trick but it does not seems to work.
It seems like we need to implement window logic properly into the Runfiles struct.

I will have to re-visit this later 🤔

**Comment by sluongng** [comment]
> Thanks for working on this, let me know if I can help.
> 
> Just one comment for now: It would be great if we could make the runfiles library available at the canonical location as per https://docs.bazel.build/versions/main/skylark/deploying.html#runfiles-library, even if only via an alias.

Yeah I think alias would be the way to do it.

That might be 2-3 PR into the future though. As I noted in the description, I plan to visit how the current API work after this and refactor / deprecate things where possible.

File: go/tools/bazel/runfiles/fs_test.go
Comment (sluongng) [inline go/tools/bazel/runfiles/fs_test.go:67]: Looking for help to make this work for windows.

File: go/tools/bazel/runfiles/runfiles_test.go
Comment (sluongng) [inline go/tools/bazel/runfiles/runfiles_test.go:67]: Needs help making this work for windows.

**Review by sluongng — COMMENTED** [review]
Small updates:

After several iteration in https://github.com/bazelbuild/rules_go/pull/3209 I could not figure out how to make this windows compatible. I don't have a local window environment to work with. In addition to that, some of the window behaviors was not clearly defined in java/python runfiles and the related test suites, which I was using as references for this Go's implementation.

image
https://github.com/bazelbuild/bazel/blob/884156ffc35eccb9ab56c849817cbf045986c968/tools/java/runfiles/testing/RunfilesTest.java#L246-L263

I would appreciated if somebody could help me test and potentially fix to make this work on windows. 🙏

Here are the changes from my latest iterations:
- Adjusted how 'canonical' path is evaluated. This could potentially deviate from the litterature in the GDoc proposal, but make things more consistent vs the native runfiles libs in @bazel_tools. Most notable change is that `..` and `.` are accepted under current spec.

- Beefed up the test suite by including Bash's runfiles as the external data dependency. This should help highlighted how Windows's result could be quite inconsistent to work with.

- Temporarily disabled TestFS on Window. Not yet sure what to do with `ExampleRunfiles` test.

- Replaced all usages of `path` APIs, which is unix specific, in favor of `path/filepath`.

- Minor costmetic changes in hope that it helps improve readability

**Comment by fmeum** [comment]
Thanks for digging into this! I should be able to take over and debug the Windows issues in the coming weeks.

**Comment by sluongng** [comment]
@fmeum Did you have a chance to review what is the expectations for runfiles lib on window?

**Comment by fmeum** [comment]
@sluongng I'm trying some things over at https://github.com/bazelbuild/rules_go/pull/3253. Looks like I got something wrong in my initial implementation of "runfiles in rundirs" lookup, sorry for that.

**Comment by fmeum** [comment]
@sluongng Sorry for the long wait, Bzlmod ate up most of my "Bazel budget".

Tests should be passing over at https://github.com/bazelbuild/rules_go/pull/3253 now. Feel free to take my two commits and squash them into yours.

Having the runfiles library upstream will also make it easier to implement https://github.com/bazelbuild/bazel/issues/16124 for Go, which I'm going to work on next.

**Comment by fmeum** [comment]
@sluongng If you are busy, would you be okay with me squashing my changes into your branch for this PR and merging?

**Comment by sluongng** [comment]
Yes please go ahead. I am a bit occupied for the next 2 months on non-bazel
stuffs 🙈

**Review by fmeum — APPROVED** [review]

**Comment by fmeum** [comment]
@linzhp @achew22 I reviewed the code and fixed the tests. Since is a rather large PR and I have authored some parts, I would like to get an approval from one of you before merging.

File: go/tools/bazel/runfiles/global.go
Comment (linzhp) [inline go/tools/bazel/runfiles/global.go:47]:
```suggestion
	runfiles    *Runfiles
```

**Review by linzhp — APPROVED** [review]
minor suggestion. LGTM

File: go/tools/bazel/runfiles/runfiles.go
Comment (linzhp) [inline go/tools/bazel/runfiles/runfiles.go:112]:
```suggestion
func (r *Runfiles) Path(relPath string) (string, error) {
```

Comment (fmeum) [reply go/tools/bazel/runfiles/runfiles.go:112]: The function allows absolute paths as well, so I am going to use `path` instead and adding an example.

**Comment by fmeum** [comment]
I also added a desperation comment to the old runfiles function. It's package pro ides other functions for which we don't have full replacements yet, 0ut that particular can be replaced.

**Comment by fmeum** [comment]
Thanks again, @sluongng!

**Pull request merged**
